seed/utils/match.py

```diff
@@ -0,0 +1,155 @@
+    """
+    empty_state = StateClass()
+    return matching_filter_criteria(
+        organization_id,
+        StateClass.__name__,
+        empty_state
+    )
+
+
+def matching_filter_criteria(organization_id, table_name, state):
+    """
+    For a given -State, returns a dictionary of it's matching criteria values.
+
+    This dictionary is frequently unpacked for a QuerySet filter or exclude.
+    """
+    return {
+        column_name: getattr(state, column_name, None)
+        for column_name
+        in matching_criteria_column_names(organization_id, table_name)
+    }
+
+
+def matching_criteria_column_names(organization_id, table_name):
+    """
+    Collect matching criteria columns while replacing address_line_1 with
+    normalized_address if applicable. A Python set is returned to handle the
+    case where normalized_address might show up twice, which shouldn't really
+    happen anyway.
+    """
+    return {
+        'normalized_address' if c.column_name == "address_line_1" else c.column_name
+        for c
+        in Column.objects.filter(
+            organization_id=organization_id,
+            is_matching_criteria=True,
+            table_name=table_name
+        )
+    }
+
+
+def match_merge_in_cycle(view_id, StateClassName):
+    """
+    Given a -View ID, this method matches and merges for the related -State.
+    Match-eligible -States are scoped to those associated with -Views within
+    the same Cycle.
+
+    If the -State associated with the -View doesn't have any matching criteria
+    values populated, the -State is not eligible for a match merge.
+    """
+    if StateClassName == 'PropertyState':
+        StateClass = PropertyState
+        ViewClass = PropertyView
+    elif StateClassName == 'TaxLotState':
+        StateClass = TaxLotState
+        ViewClass = TaxLotView
+
+    view = ViewClass.objects.get(pk=view_id)
+    org_id = view.state.organization_id
+
+    # Check if associated -State has empty matching criteria.
+    if StateClass.objects.filter(pk=view.state_id, **empty_criteria_filter(org_id, StateClass)).exists():
+        return 0, None
+
+    matching_criteria = matching_filter_criteria(org_id, StateClassName, view.state)
+    views_in_cycle = ViewClass.objects.filter(cycle_id=view.cycle_id)
+    state_matches = StateClass.objects.filter(
+        pk__in=Subquery(views_in_cycle.values('state_id')),
+        **matching_criteria
+    ).exclude(pk=view.state_id)
+
+    state_ids = list(state_matches.order_by('id').values_list('id', flat=True))
+    state_ids.append(view.state_id)  # Excluded above and appended to give merge precedence
+    count = len(state_ids)
+
+    if count > 1:
+        merged_state = merge_states_with_views(state_ids, org_id, 'System Match', StateClass)
+        view_id = ViewClass.objects.get(state_id=merged_state.id).id
+        return count, view_id
+    elif count == 1:
+        return 0, None
+
+
+def whole_org_match_merge(org_id):
```

Comment:
Note, this method and tests will be heavily changed in year-over-year work. It might not be the most useful to review now.